apps/src/lib/node/ledger/shell/mod.rs

```diff
@@ -167,12 +169,55 @@ pub(super) enum ShellMode {
     Validator {
         data: ValidatorData,
         broadcast_sender: UnboundedSender<Vec<u8>>,
-        ethereum_recv: UnboundedReceiver<EthereumEvent>,
+        ethereum_recv: EthereumReceiver,
     },
     Full,
     Seed,
 }
 
+/// A channel for pulling events from the Ethereum oracle
+/// and queueing them up for inclusion in vote extensions
+#[derive(Debug)]
+pub(super) struct EthereumReceiver {
+    channel: UnboundedReceiver<EthereumEvent>,
+    queue: Vec<EthereumEvent>,
+}
+
+impl EthereumReceiver {
+    /// Create a new [`EthereumReceiver`] from a channel connected
+    /// to an Ethereum oracle
+    pub fn new(channel: UnboundedReceiver<EthereumEvent>) -> Self {
+        Self {
+            channel,
+            queue: vec![],
+        }
+    }
+
+    /// Pull messages from the channel and add to queue
+    /// Since vote extensions require ordering of ethereum
+    /// events, we do that here. We also de-duplicate events
+    pub fn fill_queue(&mut self) {
+        while let Ok(eth_event) = self.channel.try_recv() {
+            self.queue.push(eth_event);
```

Comment:
[`BTreeSet`](https://doc.rust-lang.org/stable/std/collections/struct.BTreeSet.html) should keep everything sorted and de-duplicated.